### PR TITLE
Centralized metadata table 

### PR DESCRIPTION
**Pull request by veloman-yunkan**
This PR addresses https://github.com/openzim/zim-tools/issues/334#issuecomment-1476307896

### CODE REVIEW

File: src/metadata.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+const Metadata::ReservedMetadataTable Metadata::reservedMetadataInfo = {
+  // name              isMandatory   minLength   maxLength   regex
+  { "Name",            true, 	      /*???*/10,  /*???*/99,  /*???*/""         },
+  { "Title",           true, 	      /*???*/10,  /*???*/30,  /*???*/""         },
+  { "Creator",         true, 	      /*???*/10,  /*???*/30,  /*???*/""         },
+  { "Publisher",       true, 	      /*???*/10,  /*???*/30,  /*???*/""         },
+  { "Date",            true, 	      10,         10,         "YYYY-MM-DD"      },
+  { "Description",     true,        10,         80,         ""                },
+  { "LongDescription", false,       10,         4000,       ""                },
```
[the remark this answers is not in the document]
Comment (holta) [reply src/metadata.cpp:33]: Thx for explaining!

Is https://wiki.openzim.org/wiki/Metadata considered the more "official" spec, as it evolves?

**Comment by veloman-yunkan** [comment]
Inviting @mgautierfr @kelson42 to review the overall approach for centralizing all metadata checks in one place. Please look at the file `metadata_constraints.cpp` (note that it is not compiled separately, but is rather `#include`d in `metadata.cpp`. 

Issues noticed so far:

- regex may be a convenient tool as a lowest common denominator for validating various strings, however it falls short for some types of data if we need more rigorous checks. In our case, validating the date with a simple regex will allow entering invalid dates (e.g. 30th of February). This can be fixed by changing the last field in the metadata info table from a regexp to a special polymorphic validator object - a concrete implementation of thereof can be created from a regular expression. Do we want to do it now?
- complex checks involving dependencies between different metadata entries (e.g. the long description, if present, should be longer than the short description; the text of description metadata should be in the language matching the value of the Language metadata, etc). A draft implementation for such checks is provided in the second commit.

**Comment by kelson42** [comment]
> regex may be a convenient tool as a lowest common denominator for validating various strings, however it falls short for some types of data if we need more rigorous checks. In our case, validating the date with a simple regex will allow entering invalid dates (e.g. 30th of February). This can be fixed by changing the last field in the metadata info table from a regexp to a special polymorphic validator object - a concrete implementation of thereof can be created from a regular expression. Do we want to do it now?

@veloman-yunkan Thank you very much for your PR which is, at a first sight, really much the kind of code I was expected.

Indeed regexs fall short in many cases. The case I have in mind is checking illustration entries (requires libpng).

I'm not in favour of extending much the validation scenarios because this can go endlessly and my goal is to have the basic ones.

But I would appreciate to have the appropriate code infrastructure so we can easily do that work later... without rearchitecturing effort.

**Comment by kelson42** [comment]
On my local computer, it does not compile:
```
FAILED: test/metadata-test.p/.._src_metadata.cpp.o 
ccache c++ -Itest/metadata-test.p -Itest -I../test -I/usr/local/include -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++11 -g -Werror -Wall '-DVERSION="3.1.3"' -DGTEST_HAS_PTHREAD=1 -pthread -MD -MQ test/metadata-test.p/.._src_metadata.cpp.o -MF test/metadata-test.p/.._src_metadata.cpp.o.d -o test/metadata-test.p/.._src_metadata.cpp.o -c ../src/metadata.cpp
../src/metadata.cpp: In static member function ‘static const zim::Metadata::ReservedMetadataRecord& zim::Metadata::getReservedMetadataRecord(const string&)’:
../src/metadata.cpp:100:14: error: ‘out_of_range’ is not a member of ‘std’
  100 |   throw std::out_of_range(name + " is not a reserved metadata name");
      |              ^~~~~~~~~~~~
../src/metadata.cpp:101:1: error: control reaches end of non-void function [-Werror=return-type]
  101 | }
      | ^
cc1plus: all warnings being treated as errors
[41/48] Compiling C++ object test/tools-test.p/.._src_zimwriterfs_zimcreatorfs.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by rgaudin** [comment]
I have mixed feelings about this. On one hand, this mainly highlights the shortcomings of such an approach but on the other hand, simple checks are better than none.

Couple of comments (already identified):
- I don't think a language regexp makes sense. If you want to ensure validity, you can only work of a valid list IMO. Making sure that it's 3 lowercase letters (or proper list of such) would already be beneficial.
- date regexp wont capture dates such as 2023-13-01. Going the extra mile (as suggested) would not guarantee a valid date  with cases like Feb 29th… that would require a date dependency
- min length for all those string metadata looks problematic. Max length as well as long as the spec is silent about it. We could raise a warning above a certain threshold though.
- I am not a fan of the “LongDescription shouldn't be shorter than Description” because that's not in the spec.
- “Description must be in the language of the ZIM file” is an interesting than deserves a discussion and decision. I am in favor of updating the spec on this to specify whatever we decide.
- Illustrations are not checked despite it being the ticket's moto.

Once again we fall short on setting clear goals for our tools. zimcheck's description is “zimcheck checks the quality of a ZIM file.”. Does that mean that whenever zimcheck doesn't report an issue, the ZIM is guaranteed to be valid?

I join @kelson42 in thinking we want basic checks for now that we could extend in the future.
- Simple date format regex
- Simple format regexp for language
- [magic number](https://gist.github.com/leommoore/f9e57ba2aa4bf197ebc5) check for Illustrations
- len >= 1 for mandatory texts
- len <= max for bounded texts

And that's it. The rest can be discussed and extended in separate tickets, raised by actual needs.

Although it serves a different purpose, scraperlib now (not being used yet) enforces correct metadata with more elaborate checks (actual language code, proper PNG with correct sizes, etc) so most of what we produce shall be valid in this regard.

**Comment by holta** [comment]
Very thoughtful response from @rgaudin, and a big thank you to all 🏗️ 

I strongly support organic and free-form metadata standards (what's needed are strong norms and strong guidelines *not* bureaucratic rules) that allow grassroots initiatives to collaborate & innovate efficiently.

In fact even semi-structured data sometimes has an extremely valuable place along the way — thereby empowering regional and specialized communities to build their own ZIM files, with the metadata that their region/profession/culture truly needs.

For this reason I very strongly support allowing "free-form metadata fields" that not only permit but encourage grassroots (not centralized) community innovation to truly flourish.

Then later on, as strong community norms are independently nurtured + demonstrated + proven year-by-year-by-year, the world should honor those great grassroots practices — as they become more official metadata standards.

Central authorities (Kiwix) should provide basic guard rails & guidelines if course, but that's sufficient 👍 

Thank you to *everyone* including @veloman-yunkan and @kelson42 and @mgautierfr working very hard and this *critical question* helping it the evolve quickly in coming years, and every step of the way.

**Comment by mgautierfr** [comment]
I have open a issue #340 to discuss what to test and how. Please continue the discussion there and keep this discussion for PR review.

**Comment by kelson42** [comment]
I just want to emphasis again that we need this basic metadata feature pretty quickly. Thank you for having open sode tickets to continue the interesting discussion and keep this PR "simple" without "insulting" the future.

File: src/metadata.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (holta) [reply src/metadata.cpp]: > Is https://wiki.openzim.org/wiki/Metadata considered the more "official" spec, as it evolves?

Is my question going to answered, or will my comments continue to be removed?

[the remark this answers is not in the document]
Comment (kelson42) [reply src/metadata.cpp]: I don’t understand the question which seems somehow loaded…. « More official » than what? To know what is the purpose of the wiki.openzim.org pages, its enough to read them, « Metadata » article included.

[the remark this answers is not in the document]
Comment (holta) [reply src/metadata.cpp]: @mgaudin talks about the "spec" which is a step forward I agree.

So I'm just trying to understand what spec is being talked about.

If people are unwilling to answer the question in plain English (perhaps we have a translation problem) what I will do going forward now is email to ask him directly, as my questions are obviously no longer welcome here.

[the remark this answers is not in the document]
Comment (kelson42) [reply src/metadata.cpp]: It’s true that PR discussions are targetting developers and the ticket requester to a certain extend. As such it’s usually a bit hard to intervain if you don’t belong to one of these categories. Github issues and discussions are far more open… and slack channels are even less regulated. With emails, you really reduce the chance to get an answer at all, as we prefer to have open discussion, so it can be archived.

**Comment by kelson42** [comment]
@veloman-yunkan I hope you got all the feedbacks needed, let us know othwerwise.

File: src/metadata.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (holta) [reply src/metadata.cpp]: > question which seems somehow loaded

> you don’t belong

Insulting and infantilizing me for asking an honest question was really not necessary.

I honestly don't know why this forum has deteriorated so dramatically (all my questions removed or hidden).

In any case I'm tired of being bullied.

So I will use private channels where I'm treated with respect.

**Comment by veloman-yunkan** [comment]
With illustration metadata in the mix, in what units min and max length values should be specified - bytes or (unicode) characters?

1. Most metadata being textual, it's better to specify the metadata size limits in characters and not impose any size constraints on illustration metadata.
1.5 Like above, but handle illustration metadata as a special case in the C++ code and interpret its size limits in bytes.
2. Specify size limits in bytes. Then languages with non-latin scripts will be penalized with lower human-comprehensible upper limits on the length of the description metadata.

**Comment by veloman-yunkan** [comment]
A preliminary version of this PR (reflecting feedback received on the draft version) is ready. Note that the PR doesn't utilize the new `Metadata` utility in any of the zim tools (zimcheck, zimwriterfs).

**Comment by kelson42** [comment]
@veloman-yunkan Thank you, just to ensure we understand us well, this is mandatory that this is used at least in zimcheck.

**Comment by kelson42** [comment]
@mgautierfr any feedback/review?

File: src/metadata.cpp [hunk 2]
```diff
@@ -47,6 +47,59 @@ size_t getTextLength(const std::string& utf8EncodedString)
...
+  }
+
+  virtual ~MetadataComplexCheckBase() = default;
+
+  virtual bool checkMetadata(const Metadata& m) const = 0;
+
+  static const MetadataComplexCheckBase* getLastCheck() { return last; }
+
+private: // functions
+  static const MetadataComplexCheckBase* last;
```
Comment (mgautierfr) [inline src/metadata.cpp:77]: Keeping a static pointer to `this` is risky as the object can be moved. 
We only create global object for now but it is still tricky.
If we do that, it would be better to at least delete copy/move constructor and assignment operator.

**Review by mgautierfr — CHANGES_REQUESTED** [review]
Seems globally good to me.
A small comment on the static pointer thing but the global structure is good.

How do you plan to implement `checkTextLanguage` ? 
Or it is just a sample and it have to be removed before merge ?

**Comment by mgautierfr** [comment]
> In this PR checkTextLanguage() was introduced only to illustrate support for complex constraints. I will get rid of it before merging.

Ok. You can remove it right now (or comment it to not loose the illustration purpose) while you fix the issue with `last` pointer.

**Comment by veloman-yunkan** [comment]
> @veloman-yunkan Thank you, just to ensure we understand us well, this is mandatory that this is used at least in zimcheck.

I added usage of this new utility in `zimwriterfs`. As a result a few issues (including two preexisting bugs related to the `--longDescription` option of `zimwriterfs) were uncovered during testing. I will finalize this PR tomorrow.

File: src/metadata.cpp [hunk 3]
```diff
@@ -112,6 +116,21 @@ bool CLSNAME::checkMetadata(const Metadata& data) const        \
...
+    if (std::isprint(c)) {
+      os << c;
+    } else {
+      const unsigned int charVal = static_cast<unsigned char>(c);
+      os << "\\x" << std::setw(2) << std::setfill('0') << charVal;
+    }
+  }
+  return os.str();
+}
+
```
Comment (mgautierfr) [inline src/metadata.cpp:137]: I think this can be moved in `tools.(cpp|h)`

**Review by mgautierfr — CHANGES_REQUESTED** [review]
> Note that metadata constraints were relaxed in order to let the zimcheck
unittests pass without updating/regenerating the unittest ZIM data.

How complex it is to regenerate the unittest data ?

File: src/metadata.cpp (hunk 3 above)
Comment (veloman-yunkan) [reply src/metadata.cpp:137]: I was about to move it, but then noticed that it will result in some confusion due to the presence of the functions `urlEncode()` and `needsEscape()` in an unnamed namespace in `tools.cpp` - within that translation unit the verb `escape` will denote two different ways of escaping a char in a string.

**Comment by veloman-yunkan** [comment]
> > Note that metadata constraints were relaxed in order to let the zimcheck
> > unittests pass without updating/regenerating the unittest ZIM data.
> 
> How complex it is to regenerate the unittest data ?

The regeneration itself should be easy. The question is whether I should use the current release of `zimwriterfs` or the one used to create the current version of test ZIM files.

**Comment by mgautierfr** [comment]
> The regeneration itself should be easy. The question is whether I should use the current release of zimwriterfs or the one used to create the current version of test ZIM files.

I think here this is not relevant (at least for the change we want to made).
But the question is not easy to answer for generic case. The solution is probably in creating a full test suite with zim files created/tested with different version (https://github.com/openzim/zim-testing-suite)

File: src/metadata.cpp (hunk 3 above)
Comment (mgautierfr) [reply src/metadata.cpp:137]: Ok. Let's move on on this. If we need the function elsewhere, we will think about its renaming (or not)

**Comment by veloman-yunkan** [comment]
The PR is final (though needs at least a rebase&fixup unless other issues are spotted).

Known issues:

- Regexp constraints can duplicate length constraints. For example for Language metadata the following pair of errors is reported when the language is specified using a 2-char code:
    - `Language must contain at least 3 characters`
    - `Language doesn't match regex: \w{3}(,\w{3})*`

- Regexp errors are not user-friendly. A solution to this issue is to provide a human comprehensible description of the constraint along the regexp. This approach smoothly extends to the [proposed more general constraint framework with polymorphic validators (with regexps being only one of implementations)](https://github.com/openzim/zim-tools/pull/339#issuecomment-1482964168).

**Comment by kelson42** [comment]
@veloman-yunkan For now, I'm happy with this kind of regex error.

**Review by mgautierfr — APPROVED** [review]

**Comment by mgautierfr** [comment]
Please rebase-fixup and we can merge.
